File: src/semantic_tracer/resolver.py
"""Resolve inventory realization candidates under a named evidence policy.

Consumes precomputed evidence packets (design spec 0003); it never executes
a realization, an operation, or a conformance recipe itself. Zero eligible
candidates rejects; more than one eligible candidate rejects as ambiguous
rather than selecting silently by lexical or load order (design spec 0001).

Must not import the conformance runner (`evidence.py`), the operation
registry (`operations.py`), domain semantics (`domain.py`), or the execution
module (`execution.py`).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from semantic_tracer.canonical import content_identity
from semantic_tracer.explanation import ExplanationNode
from semantic_tracer.jsonutil import DocumentError, require_key, require_object, require_str
from semantic_tracer.packets import EvidenceResultPacket, ProducerDiagnostic, ProducerOutcome
from semantic_tracer.realization import Realization
from semantic_tracer.reasons import (
    REASON_AMBIGUOUS,
    REASON_ASSUMPTIONS_NOT_ALLOWED,
    REASON_CATEGORY_NOT_ACCEPTED,
    REASON_CONFORMANCE_FAILED,
    REASON_MISSING_EVIDENCE,
    REASON_NO_ELIGIBLE,
    REASON_OBLIGATION_NOT_GOVERNED,
    REASON_OBLIGATION_SET_UNSUPPORTED,
    REASON_THEORY_MISMATCH,
)
from semantic_tracer.theory import Theory, required_obligation_id
from semantic_tracer.types import JsonObject

CLAIM_ARTIFACT_KIND = "resolution_claim"
CLAIM_SCHEMA_VERSION = 1

CHANGE_OPTIONS = {
    REASON_MISSING_EVIDENCE: "Add one matching conformance suite for the required obligation.",
    REASON_CATEGORY_NOT_ACCEPTED: "Supply evidence in an accepted category or change the policy.",
    REASON_ASSUMPTIONS_NOT_ALLOWED: "Remove the assumptions or use a policy that permits them.",
    REASON_CONFORMANCE_FAILED: "Fix the realization or explicitly revise the frozen contract.",
    REASON_OBLIGATION_NOT_GOVERNED: "Add an explicit policy rule for the theory obligation.",
    REASON_THEORY_MISMATCH: "Target the exact authored theory identifier.",
    "ambiguous_evidence": "Retain exactly one evidence result for the theory and obligation.",
    "evidence_obligation_mismatch": "Bind the suite to the obligation declared by the theory.",
    REASON_OBLIGATION_SET_UNSUPPORTED: (
        "Use the single-obligation v0 contract or extend the resolver."
    ),
    "stale_evidence_recipe": "Re-author the suite against the exact normalized theory identity.",
    "unbound_operation": "Bind every required operation to an available execution adapter.",
}


@dataclass(frozen=True, slots=True)
class Candidate:
    realization: Realization
    eligible: bool
    reason_codes: tuple[str, ...]
    evidence: EvidenceResultPacket | None
    diagnostic: ProducerDiagnostic | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "realization_id": self.realization.realization_id,
            "realization_identity": self.realization.identity,
            "targets_theory": self.realization.targets_theory,
            "eligible": self.eligible,
            "reason_codes": list(self.reason_codes),
            "evidence": self.evidence.to_dict() if self.evidence is not None else None,
            "producer_diagnostic": self.diagnostic.to_dict()
            if self.diagnostic is not None
            else None,
            "counterexamples": (
                list(self.evidence.counterexamples) if self.evidence is not None else []
            ),
        }

    def explanation(self) -> ExplanationNode:
        children: tuple[ExplanationNode, ...] = ()
        if self.evidence is not None:
            evidence = self.evidence
            evidence_node = ExplanationNode(
                rule="evaluate_conformance_evidence",
                outcome="passed" if evidence.passed else "failed",
                subject=evidence.realization_identity,
                details={
                    "category": evidence.category,
                    "passed_cases": evidence.passed_cases,
                    "total_cases": evidence.total_cases,
                    "assumptions": list(evidence.assumptions),
                    "counterexamples": list(evidence.counterexamples),
                },
            )
            children = (evidence_node,)
        elif self.diagnostic is not None:
            children = (
                ExplanationNode(
                    rule="evaluate_conformance_evidence",
                    outcome="no_result",
                    subject=self.realization.identity,
                    details={
                        "reason_code": self.diagnostic.reason_code,
                        "detail": self.diagnostic.detail,
                    },
                ),
            )
        return ExplanationNode(
            rule="evaluate_realization_candidate",
            outcome="eligible" if self.eligible else "rejected",
            subject=self.realization.realization_id,
            details={
                "realization_identity": self.realization.identity,
                "reason_codes": list(self.reason_codes),
                "assumptions": list(self.realization.assumptions),
                "change_options": [
                    CHANGE_OPTIONS[reason]
                    for reason in self.reason_codes
                    if reason in CHANGE_OPTIONS
                ],
            },
            children=children,
        )


@dataclass(frozen=True, slots=True)
class Resolution:
    status: str
    selected_realization: str | None
    reason_codes: tuple[str, ...]
    candidates: tuple[Candidate, ...]

    def to_dict(self) -> dict[str, Any]:
        return {
            "status": self.status,
            "selected_realization": self.selected_realization,
            "reason_codes": list(self.reason_codes),
            "candidates": [candidate.to_dict() for candidate in self.candidates],
        }


def _evaluate_candidate(
    theory: Theory,
    realization: Realization,
    required_obligation: str | None,
    outcome: ProducerOutcome | None,
    policy: JsonObject,
) -> Candidate:
    if not realization.targets_theory:
        return Candidate(
            realization=realization,
            eligible=False,
            reason_codes=(REASON_THEORY_MISMATCH,),
            evidence=None,
        )

    if required_obligation is None:
        return Candidate(
            realization=realization,
            eligible=False,
            reason_codes=(REASON_OBLIGATION_SET_UNSUPPORTED,),
            evidence=None,
        )

    if not isinstance(outcome, EvidenceResultPacket):
        diagnostic = (
            outcome
            if isinstance(outcome, ProducerDiagnostic)
            else ProducerDiagnostic(REASON_MISSING_EVIDENCE)
        )
        return Candidate(
            realization=realization,
            eligible=False,
            reason_codes=(diagnostic.reason_code,),
            evidence=None,
            diagnostic=diagnostic,
        )

    evidence = outcome
    reasons: list[str] = []
    requirements = require_object(
        require_key(policy, "requirements", "policy"), "policy.requirements"
    )
    requirement = requirements.get(evidence.obligation)
    if requirement is None:
        reasons.append(REASON_OBLIGATION_NOT_GOVERNED)
    else:
        requirement_object = require_object(
            requirement, f"policy.requirements.{evidence.obligation}"
        )
        accepted_categories = requirement_object.get("accepted_categories", [])
        category_accepted = (
            isinstance(accepted_categories, list) and evidence.category in accepted_categories
        )
        if not category_accepted:
            reasons.append(REASON_CATEGORY_NOT_ACCEPTED)
        allow_assumptions = requirement_object.get("allow_assumptions") is True
        assumptions_present = bool(evidence.assumptions) or bool(realization.assumptions)
        if assumptions_present and not allow_assumptions:
            reasons.append(REASON_ASSUMPTIONS_NOT_ALLOWED)

    if not evidence.passed:
        reasons.append(REASON_CONFORMANCE_FAILED)

    return Candidate(
        realization=realization,
        eligible=not reasons,
        reason_codes=tuple(reasons),
        evidence=evidence,
    )


def resolve(
    theory: Theory,
    theory_id: str,
    realizations: list[Realization],
    evidence_outcomes: dict[str, ProducerOutcome],
    policy: JsonObject,
) -> Resolution:
    ambiguity = require_str(require_key(policy, "ambiguity", "policy"), "policy.ambiguity")
    if ambiguity != "reject":
        raise DocumentError(f"unsupported ambiguity policy {ambiguity!r}")

    required_obligation = required_obligation_id(theory)
    candidates = tuple(
        _evaluate_candidate(
            theory,
            realization,
            required_obligation,
            evidence_outcomes.get(realization.realization_id),
            policy,
        )
        for realization in realizations
    )
    eligible = [candidate for candidate in candidates if candidate.eligible]

    if len(eligible) == 1:
        selected = eligible[0]
        return Resolution(
            status="selected",
            selected_realization=selected.realization.realization_id,
            reason_codes=(),
            candidates=candidates,
        )
    if len(eligible) == 0:
        return Resolution(
            status="rejected",
            selected_realization=None,
            reason_codes=(REASON_NO_ELIGIBLE,),
            candidates=candidates,
        )
    return Resolution(
        status="rejected",
        selected_realization=None,
        reason_codes=(REASON_AMBIGUOUS,),
        candidates=candidates,
    )


def build_resolution_claim(
    theory: Theory,
    theory_id: str,
    policy: JsonObject,
    resolution: Resolution,
    selected_assumptions: tuple[str, ...],
) -> JsonObject:
    """Serialize `resolution` as a `resolution_claim_v1` document."""
    policy_id = require_str(require_key(policy, "id", "policy"), "policy.id")

    selected: JsonObject | None = None
    if resolution.status == "selected":
        selected_candidate = next(
            candidate
            for candidate in resolution.candidates
            if candidate.realization.realization_id == resolution.selected_realization
        )
        selected = {
            "id": selected_candidate.realization.realization_id,
            "identity": selected_candidate.realization.identity,
        }

    return {
        "artifact_kind": CLAIM_ARTIFACT_KIND,
        "schema_version": CLAIM_SCHEMA_VERSION,
        "theory": {"id": theory_id, "identity": theory.identity},
        "required_obligation": required_obligation_id(theory),
        "policy": {"id": policy_id, "content_identity": content_identity(policy)},
        "candidates": [
            {
                "realization_id": candidate.realization.realization_id,
                "realization_identity": candidate.realization.identity,
                "targets_theory": candidate.realization.targets_theory,
                "realization_assumptions": list(candidate.realization.assumptions),
                "evidence": candidate.evidence.to_dict()
                if candidate.evidence is not None
                else None,
                "producer_diagnostic": (
                    candidate.diagnostic.to_dict() if candidate.diagnostic is not None else None
                ),
                "eligible": candidate.eligible,
                "reason_codes": list(candidate.reason_codes),
            }
            for candidate in resolution.candidates
        ],
        "status": resolution.status,
        "selected": selected,
        "selected_assumptions": list(selected_assumptions),
    }

File: src/tracer/resolver.ts
import {
  evidenceCounterexamples,
  evidencePassed,
  evidencePassedCases,
  evidenceToJson,
  evidenceTotalCases,
  producerDiagnosticToJson,
  type EvidenceResult,
  type ProducerDiagnostic,
  type ProducerDiagnosticKind,
  type ProducerOutcome,
} from "./evidence.ts";
import type { ExplanationNode } from "./explanation.ts";
import {
  DocumentError,
  requireKey,
  requireObject,
  requireString,
  type JsonObject,
} from "./json.ts";
import { realizationAssumptions, realizationId, type Realization } from "./realization.ts";
import type { Theory } from "./theory.ts";

export const REASON_MISSING_EVIDENCE = "missing_evidence";
export const REASON_CATEGORY_NOT_ACCEPTED = "evidence_category_not_accepted";
export const REASON_ASSUMPTIONS_NOT_ALLOWED = "assumptions_not_allowed";
export const REASON_CONFORMANCE_FAILED = "conformance_failed";
export const REASON_OBLIGATION_NOT_GOVERNED = "obligation_not_governed";
export const REASON_AMBIGUOUS = "ambiguous_candidates";
export const REASON_NO_ELIGIBLE = "no_eligible_candidates";
export const REASON_THEORY_MISMATCH = "theory_mismatch";
export const REASON_EVIDENCE_AMBIGUOUS = "ambiguous_evidence";
export const REASON_EVIDENCE_OBLIGATION_MISMATCH = "evidence_obligation_mismatch";
export const REASON_OBLIGATION_SET_UNSUPPORTED = "required_obligation_set_unsupported";
export const REASON_EVIDENCE_STALE = "stale_evidence_recipe";
export const REASON_OPERATION_UNBOUND = "unbound_operation";

const CHANGE_OPTIONS: Readonly<Record<string, string>> = {
  [REASON_MISSING_EVIDENCE]: "Add one matching conformance suite for the required obligation.",
  [REASON_CATEGORY_NOT_ACCEPTED]: "Supply evidence in an accepted category or change the policy.",
  [REASON_ASSUMPTIONS_NOT_ALLOWED]: "Remove the assumptions or use a policy that permits them.",
  [REASON_CONFORMANCE_FAILED]: "Fix the realization or explicitly revise the frozen contract.",
  [REASON_OBLIGATION_NOT_GOVERNED]: "Add an explicit policy rule for the theory obligation.",
  [REASON_THEORY_MISMATCH]: "Target the exact authored theory identifier.",
  [REASON_EVIDENCE_AMBIGUOUS]: "Retain exactly one suite for the theory and obligation.",
  [REASON_EVIDENCE_OBLIGATION_MISMATCH]: "Bind the suite to the obligation declared by the theory.",
  [REASON_OBLIGATION_SET_UNSUPPORTED]:
    "Use the single-obligation v0 contract or extend the resolver.",
  [REASON_EVIDENCE_STALE]: "Re-author the suite against the exact normalized theory identity.",
  [REASON_OPERATION_UNBOUND]: "Bind every required operation to an available execution adapter.",
};

export interface Candidate {
  readonly realization: Realization;
  readonly eligible: boolean;
  readonly reasonCodes: ReadonlyArray<string>;
  readonly evidence: EvidenceResult | null;
  readonly producerDiagnostic: ProducerDiagnostic | null;
}

export interface Resolution {
  readonly status: "selected" | "rejected";
  readonly selectedRealization: string | null;
  readonly reasonCodes: ReadonlyArray<string>;
  readonly candidates: ReadonlyArray<Candidate>;
}

export const candidateToJson = (candidate: Candidate): JsonObject => ({
  realization_id: realizationId(candidate.realization),
  realization_identity: candidate.realization.identity,
  targets_theory: candidate.realization.targetsTheory,
  eligible: candidate.eligible,
  reason_codes: candidate.reasonCodes,
  evidence: candidate.evidence === null ? null : evidenceToJson(candidate.evidence),
  counterexamples: candidate.evidence === null ? [] : evidenceCounterexamples(candidate.evidence),
  producer_diagnostic:
    candidate.producerDiagnostic === null
      ? null
      : producerDiagnosticToJson(candidate.producerDiagnostic),
});

export const candidateExplanation = (candidate: Candidate): ExplanationNode => {
  const children: Array<ExplanationNode> = [];
  if (candidate.evidence !== null) {
    const evidence = candidate.evidence;
    children.push({
      rule: "evaluate_conformance_evidence",
      outcome: evidencePassed(evidence) ? "passed" : "failed",
      subject: evidence.realizationIdentity,
      details: {
        category: evidence.category,
        passed_cases: evidencePassedCases(evidence),
        total_cases: evidenceTotalCases(evidence),
        assumptions: evidence.assumptions,
        counterexamples: evidenceCounterexamples(evidence),
      },
      children: [],
    });
  } else if (candidate.producerDiagnostic !== null) {
    children.push({
      rule: "produce_conformance_evidence",
      outcome: "diagnostic",
      subject: realizationId(candidate.realization),
      details: {
        kind: candidate.producerDiagnostic.kind,
        message: candidate.producerDiagnostic.message,
      },
      children: [],
    });
  }
  return {
    rule: "evaluate_realization_candidate",
    outcome: candidate.eligible ? "eligible" : "rejected",
    subject: realizationId(candidate.realization),
    details: {
      realization_identity: candidate.realization.identity,
      reason_codes: candidate.reasonCodes,
      assumptions: realizationAssumptions(candidate.realization),
      change_options: candidate.reasonCodes.flatMap((reason) =>
        CHANGE_OPTIONS[reason] === undefined ? [] : [CHANGE_OPTIONS[reason]],
      ),
    },
    children,
  };
};

export const resolutionToJson = (resolution: Resolution): JsonObject => ({
  status: resolution.status,
  selected_realization: resolution.selectedRealization,
  reason_codes: resolution.reasonCodes,
  candidates: resolution.candidates.map(candidateToJson),
});

export const requiredObligation = (theory: Theory): string | null => {
  const raw = theory.payload.obligations;
  if (!Array.isArray(raw) || raw.length !== 1) return null;
  const first = raw[0];
  if (first === null || typeof first !== "object" || Array.isArray(first)) return null;
  return typeof first.id === "string" ? first.id : null;
};

// Exhaustive over ProducerDiagnosticKind so a new producer diagnostic fails
// the typecheck here instead of falling through unmapped. `not_targeted` and
// `obligation_unsupported` are unreachable in practice: resolver's own
// targetsTheory/requiredObligation gates below always short-circuit before a
// producer outcome carrying either kind is consulted; the entries exist for
// defense in depth if a caller ever hand-builds an outcome.
const DIAGNOSTIC_REASON: Readonly<Record<ProducerDiagnosticKind, string>> = {
  not_targeted: REASON_THEORY_MISMATCH,
  obligation_unsupported: REASON_OBLIGATION_SET_UNSUPPORTED,
  missing_evidence: REASON_MISSING_EVIDENCE,
  ambiguous_evidence: REASON_EVIDENCE_AMBIGUOUS,
  stale_evidence_recipe: REASON_EVIDENCE_STALE,
  evidence_obligation_mismatch: REASON_EVIDENCE_OBLIGATION_MISMATCH,
  unbound_operation: REASON_OPERATION_UNBOUND,
};

/**
 * Consumes an already-produced evidence packet (or typed producer
 * diagnostic); it never selects a recipe, resolves an execution adapter, or
 * runs conformance itself. See the evidence-production boundary in
 * `evidence.ts`.
 */
const evaluateCandidate = (
  theory: Theory,
  realization: Realization,
  producerOutcome: ProducerOutcome,
  policy: JsonObject,
): Candidate => {
  const reject = (reason: string, producerDiagnostic: ProducerDiagnostic | null = null): Candidate => ({
    realization,
    eligible: false,
    reasonCodes: [reason],
    evidence: null,
    producerDiagnostic,
  });
  if (!realization.targetsTheory) return reject(REASON_THEORY_MISMATCH);
  const obligation = requiredObligation(theory);
  if (obligation === null) return reject(REASON_OBLIGATION_SET_UNSUPPORTED);

  if (!producerOutcome.ok) {
    return reject(DIAGNOSTIC_REASON[producerOutcome.diagnostic.kind], producerOutcome.diagnostic);
  }
  const evidence = producerOutcome.result;
  // The producer only returns `ok: true` after matching this exact
  // obligation (see evidence.ts), but the resolver re-checks explicitly
  // rather than trusting that internal invariant alone.
  if (evidence.obligation !== obligation) return reject(REASON_EVIDENCE_OBLIGATION_MISMATCH);

  const reasons: Array<string> = [];
  const requirements = requireObject(
    requireKey(policy, "requirements", "policy"),
    "policy.requirements",
  );
  const requirement = requirements[evidence.obligation];
  if (requirement === undefined) {
    reasons.push(REASON_OBLIGATION_NOT_GOVERNED);
  } else {
    const rule = requireObject(requirement, `policy.requirements.${evidence.obligation}`);
    const categories = rule.accepted_categories ?? [];
    if (!Array.isArray(categories) || !categories.includes(evidence.category)) {
      reasons.push(REASON_CATEGORY_NOT_ACCEPTED);
    }
    const assumptionsPresent =
      evidence.assumptions.length > 0 || realizationAssumptions(realization).length > 0;
    if (assumptionsPresent && rule.allow_assumptions !== true) {
      reasons.push(REASON_ASSUMPTIONS_NOT_ALLOWED);
    }
  }
  if (!evidencePassed(evidence)) reasons.push(REASON_CONFORMANCE_FAILED);
  return {
    realization,
    eligible: reasons.length === 0,
    reasonCodes: reasons,
    evidence,
    producerDiagnostic: null,
  };
};

/**
 * Binds each producer outcome to its realization by declared, unique
 * authored ID (`outcome.realizationId`), never by array position:
 * `evidenceOutcomes` may arrive in any order relative to `realizations`.
 * Coverage must be exact — one outcome per realization, no omission, no
 * duplicate binding, no outcome bound to an unknown realization. Every
 * outcome (success or diagnostic) must also declare the exact authored
 * realization content identity it claims, and every successful result must
 * declare the exact authored theory content identity; a mismatch on either
 * throws instead of silently misattributing evidence (a rebound pure result
 * or diagnostic must never make a different, broken realization eligible).
 */
const bindOutcomes = (
  theory: Theory,
  realizations: ReadonlyArray<Realization>,
  evidenceOutcomes: ReadonlyArray<ProducerOutcome>,
): ReadonlyMap<string, ProducerOutcome> => {
  const outcomeById = new Map<string, ProducerOutcome>();
  for (const outcome of evidenceOutcomes) {
    if (outcomeById.has(outcome.realizationId)) {
      throw new DocumentError({
        message: `duplicate evidence-production outcome bound to realization '${outcome.realizationId}'`,
      });
    }
    outcomeById.set(outcome.realizationId, outcome);
  }
  const knownIds = new Set(realizations.map((realization) => realizationId(realization)));
  for (const id of outcomeById.keys()) {
    if (!knownIds.has(id)) {
      throw new DocumentError({
        message: `evidence-production outcome bound to unknown realization '${id}'`,
      });
    }
  }
  for (const realization of realizations) {
    const id = realizationId(realization);
    const outcome = outcomeById.get(id);
    if (outcome === undefined) {
      throw new DocumentError({
        message: `missing evidence-production outcome for realization '${id}'`,
      });
    }
    if (outcome.realizationIdentity !== realization.identity) {
      throw new DocumentError({
        message: `evidence-production outcome for realization '${id}' carries a mismatched realization identity`,
      });
    }
    if (outcome.ok) {
      // The wrapper-level realizationIdentity check above only catches a
      // forged wrapper; also verify the embedded evidence artifact's own
      // claimed subject and theory, since a copied result could carry a
      // stale realizationIdentity even under a correctly rebound wrapper.
      if (outcome.result.realizationIdentity !== realization.identity) {
        throw new DocumentError({
          message: `evidence result for realization '${id}' carries a mismatched realization identity`,
        });
      }
      if (outcome.result.theoryIdentity !== theory.identity) {
        throw new DocumentError({
          message: `evidence-production outcome for realization '${id}' carries a mismatched theory identity`,
        });
      }
    }
  }
  return outcomeById;
};

export const resolve = (
  theory: Theory,
  realizations: ReadonlyArray<Realization>,
  evidenceOutcomes: ReadonlyArray<ProducerOutcome>,
  policy: JsonObject,
): Resolution => {
  const ambiguity = requireString(requireKey(policy, "ambiguity", "policy"), "policy.ambiguity");
  if (ambiguity !== "reject") {
    throw new DocumentError({ message: `unsupported ambiguity policy '${ambiguity}'` });
  }
  const outcomeById = bindOutcomes(theory, realizations, evidenceOutcomes);
  const candidates = realizations.map((realization) =>
    evaluateCandidate(theory, realization, outcomeById.get(realizationId(realization))!, policy),
  );
  const eligible = candidates.filter((candidate) => candidate.eligible);
  if (eligible.length === 1) {
    return {
      status: "selected",
      selectedRealization: realizationId(eligible[0]!.realization),
      reasonCodes: [],
      candidates,
    };
  }
  return {
    status: "rejected",
    selectedRealization: null,
    reasonCodes: [eligible.length === 0 ? REASON_NO_ELIGIBLE : REASON_AMBIGUOUS],
    candidates,
  };
};
